mogp_emulator/HistoryMatching.py

```diff
@@ -0,0 +1,650 @@
+                       Default is ``None``.
+        :type coords: ndarray
+        :param expectations: (Optional) Tuple of 3 numpy arrays or 2 numpy arrays and
+                             ``None`` of the form expected from GP predictions. The first
+                             array must hold the predicted mean at all query points, and
+                             the second array must hold the variances from the GP predictions
+                             at all query points. The third is not used, so can simply be a
+                             dummy array or ``None``. Can instead provide a GP object and the
+                             points to query to have the predictions made internally. Default
+                             is ``None``.
+        :type expectations: tuple holding 3 ndarrays
+        """
+
+        # Place-holder values for user-defined quantities that can be passed in
+        # as args.
+        self.gp = None
+        self.obs = None
+        self.coords = None
+        self.expectations = None
+
+        # Place-holder or default values for optional or derived quantities.
+        self.ndim = None
+        self.ncoords = None
+        self.threshold = None
+        self.I = None
+        self.NROY = None
+        self.RO = None
+
+        # If suitable keyword arguments are provided, update the relevent
+        # variable(s)
+        if self.check_gp(gp):
+            self.set_gp(gp)
+        if self.check_obs(obs):
+            self.set_obs(obs)
+        if self.check_coords(coords):
+            self.set_coords(coords)
+        if self.check_expectations(expectations):
+            self.set_expectations(expectations)
+        if self.check_threshold(threshold):
+            self.set_threshold(threshold)
+
+        self.update()
+
+
+    def _select_expectations(self):
+        r"""
+        Determine the emulator predictions to use in HistoryMatching
+
+        This method queries the internal state of the HistoryMatching object to get
+        the expectations to use in the history matching calculation. It either
+        uses the expectations that were directly provided, or makes predictions
+        at the query points using the provided GP. If both methods are set
+        simultaneously, the method raises an exception.
+
+        :returns: Expectations to use in history matching. These are either provided
+                  explicitly, or made internally using the GP and query points.
+        :rtype: tuple of 3 ndarrays or 2 ndarrays and None
+        """
+        # Check that we have exactly 1 valid combination of parameters
+        UseCoordGP = False
+        UseExpectations = False
+        if (self.check_coords(self.coords) and self.check_gp(self.gp)):
+            UseCoordGP = True
+        if self.check_expectations(self.expectations):
+            UseExpectations = True
+        if UseCoordGP and UseExpectations:

```

Comment:
worth checking for the case that neither are True (to avoid returning None)?

Reply:
Yes, that did slip through. Fixed and added a test to catch it.